mitmproxy/addons/command_history.py

```diff
@@ -0,0 +1,113 @@
+import collections
+import os
+import typing
+import atexit
+
+from mitmproxy import command
+from mitmproxy import ctx
+
+
+class CommandHistory:
+    def __init__(self, size: int = 300) -> None:
+        self.saved_commands: typing.Deque[str] = collections.deque(maxlen=size)
+
+        self.filtered_commands: typing.Deque[str] = collections.deque()
+        self.current_index: int = -1
+        self.filter_str: str = ''
+
+        _command_history_dir = os.path.expanduser(ctx.options.confdir)
+        if not os.path.exists(_command_history_dir):
+            os.makedirs(_command_history_dir)
+
+        _command_history_path = os.path.join(_command_history_dir, 'command_history')
+        _history_lines: typing.List[str] = []
+        if os.path.exists(_command_history_path):
+            _history_lines = open(_command_history_path, 'r').readlines()
+
+        self.command_history_file = open(_command_history_path, 'w')
+
+        for l in _history_lines:
+            self.add_command(l.strip())
+
+        atexit.register(self.cleanup)
+
+    def cleanup(self):
+        if self.command_history_file:
+            self.command_history_file.close()
+
+    @property
+    def last_filtered_index(self):
+        return len(self.filtered_commands) - 1
+
+    @command.command("command_history.clear")
+    def clear_history(self):
+        self.saved_commands.clear()
+        self.filtered_commands.clear()
+        self.command_history_file.truncate(0)
+        self.command_history_file.seek(0)
+        self.command_history_file.flush()
+        self.restart()
+
+    @command.command("command_history.cancel")
+    def restart(self) -> None:
+        self.filtered_commands = self.saved_commands.copy()
+        self.current_index = -1
+
+    @command.command("command_history.next")
+    def get_next(self) -> str:
+
+        if self.current_index == -1 or self.current_index == self.last_filtered_index:
+            self.current_index = -1
+            return ''
+        elif self.current_index < self.last_filtered_index:
+            self.current_index += 1
+
+        ret = self.filtered_commands[self.current_index]
+
+        return ret
```

Comment:
Hm... Do we want this as part of the addon, or do we want the addon to just provide a way to give us the entire list? The filtering etc. provides a lot of statefulness, which would be nice to avoid. Thinking ahead, if we integrate commands into mitmweb, we probably just want a single command that gives us the list (and no filtering on the Python side).

Reply:
This just filters the contents of the commands available. If you just want to use the whole list, just don't set a filter. `self.filtered` commands will have all available commands if the filter is empty. This allows for you viewing commands that were type before that start with something you know, for example, if you want to see the last `set` command you entered, just type `:set<up>` and you will get the last `set` command. Pretty useful, and mitmweb can make use of it if it wants to.